overlord/hookstate/context.go

```diff
@@ -56,13 +57,45 @@ func NewContext(task *state.Task, setup *HookSetup, handler Handler) (*Context,
 
 	return &Context{
 		task:    task,
+		state:   task.State(),
 		setup:   setup,
 		id:      base64.URLEncoding.EncodeToString(idBytes),
 		handler: handler,
 		cache:   make(map[interface{}]interface{}),
 	}, nil
 }
 
+// NewSnapContext returns a new snap Context.
+func NewSnapContext(state *state.State, setup *HookSetup, handler Handler) (*Context, error) {
+	// Generate a secure, random ID for this context
+	idBytes := make([]byte, 32)
+	_, err := rand.Read(idBytes)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate context ID: %s", err)
+	}
+
+	return &Context{
+		task:    nil,
+		state:   state,
+		setup:   setup,
+		id:      base64.URLEncoding.EncodeToString(idBytes),
+		handler: handler,
+		cache:   make(map[interface{}]interface{}),
+	}, nil
+}
+
+// NewSnapContextWithID returns a new snap Context with a predefined contextID (must be base64-encoded).
+func NewSnapContextWithID(state *state.State, setup *HookSetup, handler Handler, contextID string) *Context {
```

Comment:
Similarly, we can make contextID an optional parameter on that one function. If empty, we set it.